File: typescript/examples/basic/src/migrator.ts
import * as V1 from "../dist/v1";
import * as V2 from "../dist/v2";
import * as V3 from "../dist/v3";
import { createVersionedDataHandler, type MigrationFn } from "vbare";

export function migrateV1TodoToV2(todo: V1.Todo): V2.Todo {
  return {
    id: BigInt(todo.id) as V2.TodoId,
    title: todo.title,
    status: todo.done ? V2.TodoStatus.Done : V2.TodoStatus.Open,
    createdAt: 0n as V2.u64,
    tags: [],
  } as V2.Todo;
}

export function migrateV1ToV2App(app: V1.App): V2.App {
  const todos = new Map<V2.TodoId, V2.Todo>();
  for (const t of app.todos) {
    const migrated = migrateV1TodoToV2(t);
    todos.set(migrated.id, migrated);
  }
  return {
    todos,
    settings: new Map<string, string>(),
  } as V2.App;
}

export function migrateV2TodoToV3(todo: V2.Todo): V3.Todo {
  // Convert list<string> tags to map<TagId, Tag>
  const tags = new Map<V3.TagId, V3.Tag>();
  let nextTagId = 1 as V3.TagId; // simple incremental ids
  for (const name of todo.tags) {
    const id = nextTagId as V3.TagId;
    tags.set(id, { id, name, color: null });
    nextTagId = ((nextTagId as unknown as number) + 1) as V3.TagId;
  }

  return {
    id: todo.id as unknown as V3.TodoId,
    status: (todo.status as unknown) as V3.TodoStatus,
    createdAt: todo.createdAt as unknown as V3.u64,
    priority: V3.Priority.Medium,
    assignee: { kind: V3.AssigneeKind.None, userId: null, teamId: null },
    detail: { title: todo.title, tags },
    history: [],
  } as V3.Todo;
}

export function migrateV2ToV3App(app: V2.App): V3.App {
  const todos = new Map<V3.TodoId, V3.Todo>();
  for (const [id, t] of app.todos) {
    const migrated = migrateV2TodoToV3(t);
    todos.set(id as unknown as V3.TodoId, migrated);
  }

  return {
    todos,
    config: { theme: V3.Theme.System, features: new Map<string, boolean>() },
    boards: new Map<V3.BoardId, V3.Board>(),
  } as V3.App;
}

// Set up versioned migration handler using the vbare package.
export const CURRENT_VERSION = 3 as const;

// Map migrations as fromVersion -> (data) => nextVersionData
export const migrations = new Map<number, MigrationFn<any, any>>([
  [1, (data: V1.App) => migrateV1ToV2App(data)],
  [2, (data: V2.App) => migrateV2ToV3App(data)],
]);

// For this example we use JSON for (de)serialization to drive the migration flow.
// The focus is on demonstrating the vbare migration wiring, not binary I/O.
const jsonEncoder = new TextEncoder();
const jsonDecoder = new TextDecoder();

export const APP_VERSIONED = createVersionedDataHandler<V3.App>({
  currentVersion: CURRENT_VERSION,
  migrations,
  serializeVersion: (data: V3.App) => jsonEncoder.encode(JSON.stringify(data)),
  deserializeVersion: (bytes: Uint8Array) => JSON.parse(jsonDecoder.decode(bytes)),
});

export function migrateToLatest(
  app: V1.App | V2.App | V3.App,
  fromVersion: 1 | 2 | 3,
): V3.App {
  if (fromVersion === 3) return app as V3.App;
  // Use the versioned handler to apply migrations starting from fromVersion.
  const bytes = jsonEncoder.encode(JSON.stringify(app));
  return APP_VERSIONED.deserialize(bytes, fromVersion);
}

File: rust/examples/basic/tests/migrator.rs
use std::collections::HashMap;

use basic::{schemas, AppVersioned};
use vbare::OwnedVersionedData;

#[test]
fn migrates_v1_to_v3() {
    let app_v1 = schemas::v1::App {
        todos: vec![
            schemas::v1::Todo {
                id: 1,
                title: "a".into(),
                done: false,
            },
            schemas::v1::Todo {
                id: 2,
                title: "b".into(),
                done: true,
            },
        ],
    };

    let bytes = serde_bare::to_vec(&app_v1).unwrap();
    let migrated = AppVersioned::deserialize(&bytes, 1).unwrap();

    assert_eq!(migrated.todos.len(), 2);
    let done = migrated.todos.get(&2).unwrap();
    assert!(matches!(done.status, schemas::v3::TodoStatus::Done));
}

#[test]
fn migrates_v2_to_v3_with_tags() {
    let mut todos: HashMap<schemas::v2::TodoId, schemas::v2::Todo> = HashMap::new();
    todos.insert(
        5,
        schemas::v2::Todo {
            id: 5,
            title: "with-tags".into(),
            status: schemas::v2::TodoStatus::Open,
            created_at: 42,
            tags: vec!["red".into(), "blue".into()],
        },
    );
    let app_v2 = schemas::v2::App {
        todos,
        settings: HashMap::new(),
    };
    let bytes = serde_bare::to_vec(&app_v2).unwrap();
    let migrated = AppVersioned::deserialize(&bytes, 2).unwrap();

    let t = migrated.todos.get(&5).unwrap();
    assert_eq!(t.detail.title, "with-tags");
    assert_eq!(t.detail.tags.len(), 2);
    assert_eq!(t.created_at, 42);
}

#[test]
fn serializes_v3_to_v1() {
    // Build a minimal v3::App with one DONE todo
    let mut todos: HashMap<schemas::v3::TodoId, schemas::v3::Todo> = HashMap::new();
    todos.insert(
        7,
        schemas::v3::Todo {
            id: 7,
            status: schemas::v3::TodoStatus::Done,
            created_at: 123,
            priority: schemas::v3::Priority::High,
            assignee: schemas::v3::Assignee {
                kind: schemas::v3::AssigneeKind::None,
                user_id: None,
                team_id: None,
            },
            detail: schemas::v3::TodoDetail {
                title: "hello".into(),
                tags: HashMap::new(),
            },
            history: Vec::new(),
        },
    );

    let app_v3 = schemas::v3::App {
        todos,
        config: schemas::v3::AppConfig {
            theme: schemas::v3::Theme::System,
            features: HashMap::new(),
        },
        boards: HashMap::new(),
    };

    // Serialize to version 1 using the migrator's serialize path
    let bytes = AppVersioned::V3(app_v3).serialize(1).unwrap();

    // Decode as v1::App and assert down-conversion
    let app_v1: schemas::v1::App = serde_bare::from_slice(&bytes).unwrap();
    assert_eq!(app_v1.todos.len(), 1);
    let t = &app_v1.todos[0];
    assert_eq!(t.id, 7u32);
    assert_eq!(t.title, "hello");
    assert!(t.done);
}
